microcosm_pubsub/matchers.py

```diff
@@ -0,0 +1,95 @@
+from dataclasses import dataclass
+from json import loads
+from typing import List, Optional, Mapping
+
+from hamcrest import not_none
+from hamcrest.library.collection.issequence_containinginorder import IsSequenceContainingInOrder
+from hamcrest.library.object.hasproperty import IsObjectWithProperty
+from hamcrest.core.helpers.wrap_matcher import wrap_matcher
+
+
+@dataclass
+class PublishedMessage:
+    """
+    A published message.
+
+    """
+    topic: str
+    message: Mapping[str, str]
+
+    @property
+    def media_type(self) -> Optional[str]:
+        return self.message.get("mediaType")
+
+    @property
+    def uri(self) -> Optional[str]:
+        return self.message.get("uri")
+
+    @classmethod
+    def from_call(cls, call) -> "PublishedMessage":
+        name, args, kwargs = call
+        topic = kwargs["TopicArn"]
+        message = loads(kwargs["Message"])
+        return cls(
+            topic=topic,
+            message=message,
+        )
+
+    @staticmethod
+    def from_mock_calls(sns_producer) -> List["PublishedMessage"]:
+        return [
+            PublishedMessage.from_call(mock_call)
+            for mock_call in sns_producer.sns_client.publish.mock_calls
+            # NB: _Call is a tuple of (name, args, kwargs)
+            if "TopicArn" in mock_call[-1] and "Message" in mock_call[-1]
+        ]
+
+
+class PublishedMessageMatcher(IsObjectWithProperty):
+    """
+    Matcher that validates a single published message.
+
+    """
+    def describe_to(self, description):
+        return description.append_text(
+            f"a message with a {self.property_name} matching ",
+        ).append_description_of(
+            self.value_matcher,
+        )
+
+
+class PublishingMatcher(IsSequenceContainingInOrder):
+    """
+    Matcher that validates a sequence of published messages.
+
+    """
+    def matches(self, sns_producer, mismatch_description=None):
+        return super().matches(
+            PublishedMessage.from_mock_calls(sns_producer),
+            mismatch_description,
+        )
+
+
+def published(*matchers):
+    return PublishingMatcher([
+        wrap_matcher(matcher)
+        for matcher in matchers
+    ])
+
+
+def published_nothing(*matchers):
+    return PublishingMatcher([])
+
+
+def has_media_type(match=None):
+    if match is None:
+        match = not_none()
+
+    return PublishedMessageMatcher("media_type", wrap_matcher(match))
+
+
+def has_uri(match=None):
+    if match is None:
+        match = not_none()
```

Comment:
funny

Reply:
I didn't think about how this would read. Oddly, it is the expected semantics: if you want to match against `published(has_uri())`, we care that the message *has-a* `uri`, but nothing else.

Reply:
I agree it's the expected semantics and the right thing to do - just odd